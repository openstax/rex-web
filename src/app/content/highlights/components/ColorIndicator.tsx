import React from 'react';
import styled, { css } from 'styled-components/macro';
import { Check } from 'styled-icons/fa-solid/Check';
import { isDefined } from '../../../guards';
import { highlightStyles } from '../../constants';

interface StyleProps {
  style: typeof highlightStyles[number];
  size?: 'small';
  shape?: 'square' | 'circle';
}

interface Props<T extends React.ComponentType | undefined = React.ComponentType> extends StyleProps {
  className?: string;
  checked?: boolean;
  component?: T extends undefined ? undefined :
    T extends React.ComponentType ? React.ReactComponentElement<T>:
    never;
}

// tslint:disable-next-line:variable-name
const CheckIcon = styled(Check)`
  display: none;
`;

// tslint:disable-next-line:variable-name
const FocusedStyle = styled.span`
  display: none;
  position: absolute;
  height: ${(props: StyleProps) => indicatorSize(props) + 0.6}rem;
  width: ${(props: StyleProps) => indicatorSize(props) + 0.6}rem;
  background-color: ${(props: {style: typeof highlightStyles[number]}) => props.style.passive};
  z-index: -1;
  left: -0.4rem;
  top: -0.4rem;
  ${(props: StyleProps) => (props.shape === 'circle' || props.shape === undefined) && css`
    border-radius: 2rem;
  `}
`;

// tslint:disable-next-line:variable-name
function Hoc<T extends React.ComponentType | undefined>(props: React.PropsWithChildren<Props<T>>) {
  const {children, style, checked, size, component, ...otherProps} = props;
  const focusedProps: StyleProps = { style, size, shape: props.shape };

  if (isDefined(component)) {
    return React.cloneElement(
      component,
      props,
      <CheckIcon key='check' />,
      children,
      <FocusedStyle {...focusedProps} />
    );
  }
  return <div {...otherProps}><CheckIcon />{children}<FocusedStyle {...focusedProps} /></div>;
}

const indicatorSize = (props: {size?: 'small'}) => props.size === 'small' ? 1.6 : 2.4;
const checkSize = (props: {size?: 'small'}) => props.size === 'small' ? 1 : 1.6;

// tslint:disable-next-line:variable-name
const ColorIndicator = styled(Hoc)`
  position: relative;
  background-color: ${(props: {style: typeof highlightStyles[number]}) => props.style.passive};
  border: 1px solid ${(props: {style: typeof highlightStyles[number]}) => props.style.focused};
  height: ${(props: Props) => indicatorSize(props)}rem;
  width: ${(props: Props) => indicatorSize(props)}rem;
  display: flex;
  justify-content: center;
  align-items: center;
  ${(props: Props) => (props.shape === 'circle' || props.shape === undefined) && css`
    border-radius: 2rem;
  `}
  overflow: visible;

  ${CheckIcon} {
    height: ${(props: Props) => checkSize(props)}rem;
    width: ${(props: Props) => checkSize(props)}rem;
    ${(props: Props) => props.checked && css`
      color: ${props.style.focused};
      display: block;
    `}
  }

  input:focus + ${FocusedStyle} {
    display: block;
  }
`;

export default ColorIndicator;
