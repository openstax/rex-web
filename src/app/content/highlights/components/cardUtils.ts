import { Highlight } from '@openstax/highlighter';
import { HighlightColorEnum, HighlightUpdateColorEnum, UpdateHighlightRequest } from '@openstax/highlighter/dist/api';
import { Element, HTMLElement, } from '@openstax/types/lib.dom';
import React from 'react';
import { findElementSelfOrParent } from '../../../domUtils';
import { isHtmlElement } from '../../../guards';
import { assertWindow, remsToPx } from '../../../utils';
import { assertDefined } from '../../../utils/assertions';
import { cardMarginBottom } from '../constants';
import { HighlightData } from '../types';

export const getHighlightOffset = (container: HTMLElement | undefined, highlight: Highlight) => {
  if (!container || !highlight.range || !highlight.range.getBoundingClientRect) {
    return;
  }

  const {top, bottom } = highlight.range.getBoundingClientRect();

  const offsetParent = container.offsetParent && findElementSelfOrParent(container.offsetParent);
  const parentOffset = offsetParent ? offsetParent.offsetTop : 0;
  const scrollOffset = assertWindow().scrollY;

  return {
    bottom: bottom - parentOffset + scrollOffset,
    top: top - parentOffset + scrollOffset,
  };
};

export const getHighlightTopOffset = (container: HTMLElement | undefined, highlight: Highlight): number | undefined => {
  const offset = getHighlightOffset(container, highlight);

  if (offset) {
    return offset.top;
  }
};

export const getHighlightBottomOffset = (
  container: HTMLElement | undefined,
  highlight: Highlight
): number | undefined => {
  const offset = getHighlightOffset(container, highlight);

  if (offset) {
    return offset.bottom;
  }
};

export const generateUpdatePayload = (
  oldData: Pick<HighlightData, 'color' | 'annotation'>,
  update: {color?: HighlightColorEnum, annotation?: string, id: string }
): {
  updatePayload: UpdateHighlightRequest,
  preUpdateData: UpdateHighlightRequest
} => {
  const oldColor = oldData.color as string as HighlightUpdateColorEnum;
  const newColor = update.color !== undefined
    ? update.color as string as HighlightUpdateColorEnum
    : oldColor;

  const updatePayload = {
    highlight: {
      annotation: update.annotation !== undefined ? update.annotation : oldData.annotation,
      color: newColor,
    },
    id: update.id,
  };

  const preUpdateData = {
    highlight: {
      annotation: oldData.annotation,
      color: oldColor,
    },
    id: update.id,
  };

  return {updatePayload, preUpdateData};
};

/**
 * Calculate positions for the cards according to the location of associated highlights in the document.
 * Positions will be adjusted to the presence of the other cards in cases like multiple highlights in the same line.
 */
const updateStackedCardsPositions = (
  highlightsElements: Highlight[],
  heights: Map<string, number>,
  getHighlightPosition: (highlight: Highlight) => { top: number, bottom: number },
  initialPositions?: Map<string, number>,
  addAditionalMarginForTheFirstCard = false,
  lastVisibleCardPosition = 0,
  lastVisibleCardHeight = 0
) => {
  const positions = initialPositions ? initialPositions : new Map<string, number>();

  for (const [index, highlight] of highlightsElements.entries()) {
    const topOffset = getHighlightPosition(highlight).top;

    const marginToAdd = index > 0 || addAditionalMarginForTheFirstCard ? remsToPx(cardMarginBottom) : 0;
    const lastVisibleCardBottom = lastVisibleCardPosition + lastVisibleCardHeight;
    const stackedTopOffset = Math.max(topOffset, lastVisibleCardBottom + marginToAdd);

    if (heights.get(highlight.id)) {
      lastVisibleCardPosition = stackedTopOffset;
      lastVisibleCardHeight = heights.get(highlight.id)!;
    }

    positions.set(highlight.id, stackedTopOffset);
  }

  return positions;
};

/**
 * Calculate how much we have to move cards to adjust their offset so we can align a card for @param highlight
 * with the corresponding highlight in the document.
 */
const getOffsetToAdjustForHighlightPosition = (
  highlight: Highlight | undefined,
  cardsPositions: Map<string, number>,
  getHighlightPosition: (highlight: Highlight) => { top: number, bottom: number }
) => {
  const position = highlight
    ? assertDefined(cardsPositions.get(highlight.id), 'internal function requested postion of unknown highlight')
    : 0;

  const topOffsetFocused = highlight && position
    ? getHighlightPosition(highlight).top
    : 0;

  return position - topOffsetFocused;
};

/**
 * Calculate positions of all cards for @param highlights
 * If @param focusedHighlight is specified then change position of the related card to align it
 * with the corresponding highlight in the document and recalculate positions of all other cards that
 * may be affected by this operation.
 */
export const updateCardsPositions = (
  focusedHighlight: Highlight | undefined,
  highlights: Highlight[],
  cardsHeights: Map<string, number>,
  getHighlightPosition: (highlight: Highlight) => { top: number, bottom: number }
) => {
  const cardsPositions = updateStackedCardsPositions(highlights, cardsHeights, getHighlightPosition);

  const offsetToAdjust = getOffsetToAdjustForHighlightPosition(focusedHighlight, cardsPositions, getHighlightPosition);

  if (!focusedHighlight || offsetToAdjust === 0) { return cardsPositions; }

  const focusedHighlightIndex = highlights.findIndex((highlight) => highlight.id === focusedHighlight.id);

  // Start processing the first part of highlights

  const highlightsUpToFocused = highlights.slice(0, focusedHighlightIndex + 1);

  // Reverse an array to start from the focused highlight card, adjust its position and continue adjusting cards above
  const reversedHighlightsUpToFocused = highlightsUpToFocused.reverse();

  for (const [index, highlight] of reversedHighlightsUpToFocused.entries()) {
    const highlightCardTopOffset = cardsPositions.get(highlight.id) as number;
    const adjustedHighlightPosition = highlightCardTopOffset - offsetToAdjust;

    // Adjust position of the card to the calculated offset
    cardsPositions.set(highlight.id, adjustedHighlightPosition);

    // Check if we have to adjust position of a card above and if not then simply break
    const nextHighlight = reversedHighlightsUpToFocused[index + 1] as Highlight | undefined;
    if (nextHighlight) {
      const nextHighlightCardTopOffset = cardsPositions.get(nextHighlight.id) as number;
      const nextHighlightCardHeight = cardsHeights.get(nextHighlight.id) as number;
      const nextHighlightCardBottomOffset = nextHighlightCardTopOffset + nextHighlightCardHeight;

      if (nextHighlightCardBottomOffset <= adjustedHighlightPosition) {
        break;
      }
    }
  }

  // Start processing the second part of highlights

  // Adjusting the position of focused highlight might cause some cards to go up which means that maybe cards
  // that were previously pushed down now have space to align with their highlights.

  const highlightsAfterFocused = highlights.slice(focusedHighlightIndex + 1);

  return updateStackedCardsPositions(
    highlightsAfterFocused,
    cardsHeights,
    getHighlightPosition,
    cardsPositions,
    true,
    cardsPositions.get(focusedHighlight.id) as number,
    cardsHeights.get(focusedHighlight.id) as number
  );
};

export const noopKeyCombinationHandler = (
  allowedContainers: Array<HTMLElement | React.RefObject<HTMLElement>>
) => (activeElement: Element): boolean => {
  const activeElementIsOutsideOfTheAllowedContainer = allowedContainers.every((container) => {
    return isHtmlElement(container)
      ? !container.contains(activeElement)
      : !container.current || !container.current.contains(activeElement);
  });
  if (activeElementIsOutsideOfTheAllowedContainer) { return true; }
  if (activeElement.nodeName === 'TEXTAREA') { return true; }
  return false;
};
