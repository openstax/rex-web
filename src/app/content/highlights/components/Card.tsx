import Highlighter, { Highlight } from '@openstax/highlighter';
import { NewHighlightSourceTypeEnum } from '@openstax/highlighter/dist/api';
import { HTMLElement } from '@openstax/types/lib.dom';
import flow from 'lodash/fp/flow';
import React from 'react';
import { connect, useSelector } from 'react-redux';
import styled from 'styled-components';
import { scrollIntoView } from '../../../domUtils';
import { useFocusIn } from '../../../reactUtils';
import { AppState, Dispatch } from '../../../types';
import { highlightStyles } from '../../constants';
import * as selectHighlights from '../../highlights/selectors';
import * as selectSearch from '../../search/selectors';
import * as selectContent from '../../selectors';
import * as contentSelect from '../../selectors';
import { stripIdVersion } from '../../utils/idUtils';
import {
  clearFocusedHighlight,
  createHighlight,
  focusHighlight,
  requestDeleteHighlight,
  setAnnotationChangesPending,
} from '../actions';
import { HighlightData } from '../types';
import { getHighlightLocationFilterForPage } from '../utils';
import { mainCardStyles } from './cardStyles';
import DisplayNote from './DisplayNote';
import EditCard from './EditCard';
import showConfirmation from './utils/showConfirmation';

export interface CardProps {
  page: ReturnType<typeof selectContent['bookAndPage']>['page'];
  book: ReturnType<typeof selectContent['bookAndPage']>['book'];
  container?: HTMLElement;
  isActive: boolean;
  isTocOpen: boolean;
  hasQuery: boolean;
  highlighter: Highlighter;
  highlight: Highlight;
  create: typeof createHighlight;
  focus: typeof focusHighlight;
  remove: typeof requestDeleteHighlight;
  blur: typeof clearFocusedHighlight;
  setAnnotationChangesPending: typeof setAnnotationChangesPending;
  data?: HighlightData;
  className: string;
  zIndex: number;
  shouldFocusCard: boolean;
  topOffset?: number;
  highlightOffsets?: { top: number, bottom: number };
  onHeightChange: (ref: React.RefObject<HTMLElement>) => void;
}

// tslint:disable-next-line:variable-name
const Card = (props: CardProps) => {
  const annotation = props.data && props.data.annotation;
  const element = React.useRef<HTMLElement>(null);
  const [editing, setEditing] = React.useState<boolean>(!annotation);
  const locationFilters = useSelector(selectHighlights.highlightLocationFilters);
  const hasUnsavedHighlight = useSelector(selectHighlights.hasUnsavedHighlight);

  const { isActive, highlight: { id }, focus } = props;

  const focusCard = React.useCallback(async() => {
    if (!isActive && (!hasUnsavedHighlight || await showConfirmation())) {
      focus(id);
    }
  }, [isActive, hasUnsavedHighlight, id, focus]);

  useFocusIn(element, true, focusCard);

  React.useEffect(() => {
    if (!props.isActive) {
      setEditing(false);
    } else {
      const firstElement = props.highlight.elements[0] as HTMLElement;
      const lastElement = props.highlight.elements[props.highlight.elements.length - 1] as HTMLElement;
      const elements = [lastElement];
      if (element.current) {
        elements.push(element.current);
      }
      scrollIntoView(firstElement, elements);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [element, props.isActive]);

  React.useEffect(() => {
    if (annotation) {
      props.highlight.elements.forEach((el) => (el as HTMLElement).classList.add('has-note'));
    } else {
      props.highlight.elements.forEach((el) => (el as HTMLElement).classList.remove('has-note'));
    }
  }, [props.highlight, annotation]);

  React.useEffect(() => {
    if (!annotation && !props.isActive) {
      props.onHeightChange({ current: null });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [annotation, props.isActive]);

  const location = React.useMemo(() => {
    return props.page && getHighlightLocationFilterForPage(locationFilters, props.page);
  }, [locationFilters, props.page]);

  const locationFilterId = location && stripIdVersion(location.id);

  const { page, book } = props;
  if (!props.highlight.range || !page || !book || !locationFilterId) {
    return null;
  }

  const onRemove = () => {
    if (props.data) {
      props.remove(props.data, {
        locationFilterId,
        pageId: page.id,
      });
    }
  };
  const style = highlightStyles.find((search) => props.data && search.label === props.data.color);

  const onCreate = (isDefaultColor: boolean) => {
    props.create({
      ...props.highlight.serialize().getApiPayload(props.highlighter, props.highlight),
      scopeId: book.id,
      sourceId: page.id,
      sourceMetadata: {bookVersion: book.version},
      sourceType: NewHighlightSourceTypeEnum.OpenstaxPage,
    }, {
      isDefaultColor,
      locationFilterId,
      pageId: page.id,
    });
  };

  const commonProps = {
    className: props.className,
    highlight: props.highlight,
    isActive: props.isActive,
    onBlur: props.blur,
    onHeightChange: props.onHeightChange,
    onRemove,
    ref: element,
    shouldFocusCard: props.shouldFocusCard,
  };

  return <div onClick={focusCard} data-testid='card'>
    {
      !editing && style && annotation ? <DisplayNote
        {...commonProps}
        style={style}
        note={annotation}
        focus={props.focus}
        onEdit={() => setEditing(true)}
      /> : <EditCard
        {...commonProps}
        locationFilterId={locationFilterId}
        hasUnsavedHighlight={hasUnsavedHighlight}
        pageId={page.id}
        onCreate={onCreate}
        setAnnotationChangesPending={props.setAnnotationChangesPending}
        onCancel={() => setEditing(false)}
        data={props.data}
      />
    }
  </div>;
};

// tslint:disable-next-line: variable-name
const StyledCard = styled(Card)`
  ${mainCardStyles}
`;

export default connect(
  (state: AppState, ownProps: {highlight: Highlight}) => ({
    ...selectContent.bookAndPage(state),
    data: selectHighlights.highlights(state).find((search) => search.id === ownProps.highlight.id),
    hasQuery: !!selectSearch.query(state),
    isActive: selectHighlights.focused(state) === ownProps.highlight.id,
    isTocOpen: contentSelect.tocOpen(state),
  }),
  (dispatch: Dispatch) => ({
    blur: flow(clearFocusedHighlight, dispatch),
    create: flow(createHighlight, dispatch),
    focus: flow(focusHighlight, dispatch),
    remove: flow(requestDeleteHighlight, dispatch),
    setAnnotationChangesPending: flow(setAnnotationChangesPending, dispatch),
  })
)(StyledCard);
