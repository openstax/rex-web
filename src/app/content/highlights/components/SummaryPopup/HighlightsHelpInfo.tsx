import * as Cookies from 'js-cookie';
import React from 'react';
import { useSelector } from 'react-redux';
import styled, { css, keyframes } from 'styled-components';
import { Times } from '@styled-icons/fa-solid/Times/Times';
import { useAnalyticsEvent } from '../../../../../helpers/analytics';
import { user as userSelector } from '../../../../auth/selectors';
import { PlainButton } from '../../../../components/Button';
import htmlMessage from '../../../../components/htmlMessage';
import { bodyCopyRegularStyle } from '../../../../components/Typography';
import theme from '../../../../theme';
import { assertWindow } from '../../../../utils';
import { disablePrint } from '../../../components/utils/disablePrint';

// This is copied from CallToActionPopup > styles.tsx
// Wher should we store this kind of functions?
const slideInFromBottom = keyframes`
  0% {
    bottom: -100%;
  }

  100% {
    bottom: 0;
  }
`;

const slideInAnimation = css`
  animation: ${800}ms ${slideInFromBottom} ease-out;
`;

// tslint:disable-next-line: variable-name
const Wrapper = styled.div`
  position: fixed;
  bottom: 0;
  left: 0;
  z-index: ${theme.zIndex.highlightsHelpInfoMobile};
  ${bodyCopyRegularStyle}
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 100%;
  height: 6rem;
  padding: 0 2rem;
  background-color: ${theme.color.neutral.formBackground};
  border: 1px solid ${theme.color.neutral.formBorder};
  ${slideInAnimation}

  @media screen and (min-width: ${theme.breakpoints.mobileBreak}em) {
    display: none;
  }

  ${disablePrint}
`;

// tslint:disable-next-line: variable-name
export const CloseIcon = styled(Times)`
  color: ${theme.color.secondary.lightGray.darkest};
  width: 1.4rem;
`;

export const cookieId = 'highlights_help_info_dissmised';
export const timeBeforeShow = 1000;

// tslint:disable-next-line: variable-name
const Message = htmlMessage(
  'i18n:toolbar:highlights:popup:help-info',
  (props) => <span {...props} />);

// tslint:disable-next-line:variable-name
const HighlightsHelpInfo = () => {
  const [show, setShow] = React.useState(false);
  const user = useSelector(userSelector);
  const trackShowHelpInfo = useAnalyticsEvent('showHelpInfo');

  const dismiss = () => {
    Cookies.set(cookieId, 'true');
    setShow(false);
  };

  React.useEffect(() => {
    setTimeout(() => {
      if (
        Boolean(Cookies.get(cookieId))
        || !assertWindow().matchMedia(theme.breakpoints.mobileQuery).matches
      ) { return; }
      setShow(true);
      trackShowHelpInfo();
    }, timeBeforeShow);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  if (!show || !user) { return null; }

  return <Wrapper data-analytics-region='Mobile MH help info'>
    <Message />
    <PlainButton onClick={dismiss} data-analytics-label='close'>
      <CloseIcon />
    </PlainButton>
  </Wrapper>;
};

export default HighlightsHelpInfo;
