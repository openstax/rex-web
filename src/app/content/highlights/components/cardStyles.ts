import { css, keyframes } from 'styled-components';
import { DropdownList } from '../../../components/Dropdown';
import theme from '../../../theme';
import { remsToEms } from '../../../utils';
import {
  contentTextWidth,
  searchResultsBarDesktopWidth,
  sidebarDesktopWidth,
} from '../../components/constants';
import { disablePrint } from '../../components/utils/disablePrint';
import { highlightStyles } from '../../constants';
import {
  cardContentMargin,
  cardFocusedContentMargin,
  cardMinWindowMargin,
  cardPadding,
  cardWidth,
} from '../constants';
import { HighlightData } from '../types';
import { CardProps } from './Card';
import { getHighlightBottomOffset } from './cardUtils';
import { WrapperProps } from './CardWrapper';
import { cardBorder } from './style';

/*
 * putting overflow hidden on a page wrapper that aligns with the window edge would
 * include the sidebar, which would break position: sticky.
 *
 * avoiding using an overflow hidden to hide cards when there is not enough space
 * means being very explicit about hiding them so they don't create a horizontal
 * scrollbar.
 *
 * in this case that means using extensive knowledge about the container widths,
 * which unfortunately means knowledge of all the sidebar widths and their state
 * too.
 *
 * consider making a helper like `styleWhenSidebarClosed` maybe `styleWhenContentWidth`
 * that has a selector to get the relevant stuff.
 */

const additionalWidthForCard = (cardWidth + cardContentMargin + cardMinWindowMargin) * 2;
const minimalWidth = contentTextWidth + additionalWidthForCard;
export const minimalWidthForCards = '(max-width: ' + remsToEms(minimalWidth) + 'em)';
export const minimalWidthForCardsWithToc = '(max-width: ' +
  remsToEms(minimalWidth + sidebarDesktopWidth) + 'em)';
export const minimalWidthForCardsWithSearchResults = '(max-width: ' +
  remsToEms(minimalWidth + searchResultsBarDesktopWidth) + 'em)';

const overlapDisplay = css`
  ${(props: CardProps) => !!props.isActive && css`
    left: unset;
    right: ${cardMinWindowMargin}rem;
    top: ${props.highlightOffsets
      ? props.highlightOffsets.bottom
      : getHighlightBottomOffset(props.container, props.highlight)}px;
  `}
  ${(props: CardProps) => !props.isActive && css`
    display: none;
  `}
`;

const rightSideDisplay = css`
  left: calc(100% - ((100% - ${contentTextWidth}rem) / 2) + ${cardContentMargin}rem);
  right: unset;
  top: ${(props: CardProps) => `${props.topOffset || getHighlightBottomOffset(props.container, props.highlight)}px;`}
  ${(props: CardProps) => !!props.isActive && css`
    left: calc(100% - ((100% - ${contentTextWidth}rem) / 2) + ${cardFocusedContentMargin}rem);
  `}
`;

const touchScreenDisplay = css`
  ${(props: CardProps) => !!props.isActive && css`
    left: 0;
    right: 0;
    bottom: 0;
    top: unset;
    position: fixed;
    padding: 0;
  `}
  ${(props: CardProps) => !props.isActive && css`
    display: none;
  `}
`;

const fadeIn = keyframes`
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
`;

const fadeInAnimation = css`
  animation: ${600}ms ${fadeIn} ease-out;
`;

export const mainCardStyles = css`
  ${(props: CardProps) => !props.isActive && (!props.data || !props.data.annotation)
    ? 'display: none;'
    : 'display: block;'}
  ${fadeInAnimation}
  position: absolute;
  padding: ${cardPadding}rem;
  ${cardBorder}
  ${rightSideDisplay}
  ${disablePrint}

  z-index: ${(props: CardProps) => props.zIndex};
  transition: opacity 0.3s, top 0.3s, left 0.3s;

  ${DropdownList} {
    z-index: 1;
  }

  ${(props: {data: HighlightData}) => {
    const data = props.data;

    if (!data || !data.color) {
      return null;
    }

    const style = highlightStyles.find((search) => search.label === props.data.color);

    if (!style) {
      return null;
    }

    return css`
      ::before {
        content: ' ';
        border-radius: 0.4rem 0 0 0.4rem;
        position: absolute;
        top: 0;
        left: 0
        bottom: 0;
        width: ${cardPadding / 2}rem;
        background-color: ${style.focused};
      }
      ${theme.breakpoints.touchDeviceQuery(css`
        ::before {
          border-radius: 0.4rem 0.4rem 0 0;
          right: 0;
          bottom: unset;
          width: unset;
          height: ${cardPadding / 2}rem;
        }
     `)}
    `;
  }}

  /* in this media query if isTocOpen is null or truthy it means that toc is open */
  ${(props: CardProps) => (props.isTocOpen === null || props.isTocOpen) && css`
    @media ${minimalWidthForCardsWithToc} {
      /* the window is too small to show note cards next to content when the toc is open */
      animation: none;
      ${overlapDisplay}
    }
  `}

  /*
    not sure if this breakpoint is necessary, because search results have same width
    (from the visual perspective, because values are different)
  */
  ${(props: CardProps) => props.hasQuery && css`
    @media ${minimalWidthForCardsWithSearchResults} {
      /* the window is too small to show note cards next to content when search is open */
      animation: none;
      ${overlapDisplay}
    }
  `}

  @media ${minimalWidthForCards} {
    /* the window is too small to show note cards next to content even without sidebars */
    animation: none;
    ${overlapDisplay}
  }

  ${theme.breakpoints.touchDeviceQuery(css`
    animation: none;
    ${touchScreenDisplay}
  `)}
`;

export const mainWrapperStyles = css`
  position: relative;
  overflow: visible;
  z-index: ${theme.zIndex.highlightInlineCard};
  transform: auto;
  transition: transform 0.3s;

  ${(props: WrapperProps) => (props.isTocOpen === null || props.isTocOpen) && css`
    @media ${minimalWidthForCardsWithToc} {
      /* override js inline styles for display of all cards while using overlap display */
      transform: none !important;
    }
  `}

  ${(props: WrapperProps) => props.hasQuery && css`
    @media (max-width: ${minimalWidthForCardsWithSearchResults}em) {
      /* override js inline styles for display of all cards while using overlap display */
      transform: none !important;
    }
  `}

  @media ${minimalWidthForCards} {
    /* override js inline styles for display of all cards while using overlap display */
    transform: none !important;
  }

  ${theme.breakpoints.mobile(css`
    /* override js inline styles for display of all cards while using overlap display */
    transform: none !important;
  `)}
`;
