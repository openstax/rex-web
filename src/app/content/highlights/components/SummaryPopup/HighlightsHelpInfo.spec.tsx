import { MediaQueryList } from '@openstax/types/lib.dom';
import * as Cookies from 'js-cookie';
import React from 'react';
import renderer from 'react-test-renderer';
import createTestStore from '../../../../../test/createTestStore';
import TestContainer from '../../../../../test/TestContainer';
import { receiveUser } from '../../../../auth/actions';
import { User } from '../../../../auth/types';
import { PlainButton } from '../../../../components/Button';
import { Store } from '../../../../types';
import HighlightsHelpInfo, { cookieId, timeBeforeShow } from './HighlightsHelpInfo';

jest.mock('js-cookie', () => ({
  ...jest.requireActual('js-cookie'),
  get: jest.fn(),
  set: jest.fn(),
}));

describe('HighlightsHelpInfo', () => {
  jest.useFakeTimers();
  let store: Store;
  let user: User;

  beforeEach(() => {
    store = createTestStore();
    user = {firstName: 'test', isNotGdprLocation: true, uuid: 'a_uuid'};
  });

  it('matches snapshot when hidden', () => {
    const component = renderer.create(<TestContainer store={store}>
      <HighlightsHelpInfo/>
    </TestContainer>);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('matches snapshot when showed', async() => {
    window!.matchMedia = () => ({matches: true}) as MediaQueryList;
    store.dispatch(receiveUser(user));

    const component = renderer.create(<TestContainer store={store}>
      <HighlightsHelpInfo/>
    </TestContainer>);

    await renderer.act(async() => {
      jest.runTimersToTime(timeBeforeShow);
    });

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('does not open if user is not logged in', async() => {
    window!.matchMedia = () => ({matches: true}) as MediaQueryList;

    const component = renderer.create(<TestContainer store={store}>
      <HighlightsHelpInfo/>
    </TestContainer>);

    await renderer.act(async() => {
      jest.runTimersToTime(timeBeforeShow);
    });

    expect(() => component.root.findByProps({ 'data-testid': 'support-link' })).toThrow();
  });

  it('does not open if not on mobile', async() => {
    window!.matchMedia = () => ({matches: false}) as MediaQueryList;

    store.dispatch(receiveUser(user));

    const component = renderer.create(<TestContainer store={store}>
      <HighlightsHelpInfo/>
    </TestContainer>);

    await renderer.act(async() => {
      jest.runTimersToTime(timeBeforeShow);
    });

    expect(() => component.root.findByProps({ 'data-testid': 'support-link' })).toThrow();
  });

  it('does not open if cookie is already set', async() => {
    window!.matchMedia = () => ({matches: true}) as MediaQueryList;

    store.dispatch(receiveUser(user));

    const spy = jest.spyOn(Cookies, 'get');
    spy.mockImplementationOnce(() => 'true' as any);

    const component = renderer.create(<TestContainer store={store}>
      <HighlightsHelpInfo/>
    </TestContainer>);

    await renderer.act(async() => {
      jest.runTimersToTime(timeBeforeShow);
    });

    expect(() => component.root.findByProps({ 'data-testid': 'support-link' })).toThrow();
  });

  it('set cookie on dismiss', async() => {
    window!.matchMedia = () => ({matches: true}) as MediaQueryList;

    store.dispatch(receiveUser(user));

    const component = renderer.create(<TestContainer store={store}>
      <HighlightsHelpInfo/>
    </TestContainer>);

    await renderer.act(async() => {
      jest.runTimersToTime(timeBeforeShow);
    });

    await renderer.act(async() => {
      const dismissButton = component.root.findByType(PlainButton);
      dismissButton.props.onClick();
    });

    expect(Cookies.set).toHaveBeenCalledWith(cookieId, 'true');
  });
});
