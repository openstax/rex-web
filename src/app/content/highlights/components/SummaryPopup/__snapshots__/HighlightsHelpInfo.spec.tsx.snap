// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HighlightsHelpInfo matches snapshot when hidden 1`] = `null`;

exports[`HighlightsHelpInfo matches snapshot when showed 1`] = `
.c2 {
  display: inline-block;
  vertical-align: -.125em;
  overflow: hidden;
}

.c1 {
  cursor: pointer;
  border: none;
  margin: 0;
  padding: 0;
  background: none;
}

.c0 {
  position: fixed;
  bottom: 0;
  left: 0;
  z-index: 120;
  color: #424242;
  font-size: 1.6rem;
  line-height: 2.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
  height: 6rem;
  padding: 0 2rem;
  background-color: #f5f5f5;
  border: 1px solid #d5d5d5;
  -webkit-animation: 800ms bNiYjl ease-out;
  animation: 800ms bNiYjl ease-out;
}

.c0 a {
  color: #027EB5;
  cursor: pointer;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0 a:hover {
  color: #0064A0;
}

.c3 {
  color: #818181;
  width: 1.4rem;
}

@media screen and (min-width:75em) {
  .c0 {
    display: none;
  }
}

@media print {
  .c0 {
    display: none;
  }
}

<div
  className="c0"
  data-analytics-region="Mobile MH help info"
>
  <span
    dangerouslySetInnerHTML={
      Object {
        "__html": "Notes and highlights cannot be edited on mobile devices. <a data-testid='support-link' data-analytics-label='support-link' target=\\"_blank\\" href=\\"https://help.openstax.org/s/article/Using-the-highlighting-and-note-taking-feature-on-mobile-or-tablet?search=highlighting\\">Read more</a>",
      }
    }
  />
  <button
    aria-label="dismiss"
    className="c1"
    data-analytics-label="close"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="c2 c3"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 352 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;
